File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            multiplier = 11
            if item.name != "Aged Brie" and item.name != "Backstage passes to a TAFKAL80ETC concert":
                if item.name != "Sulfuras, Hand of Ragnaros":
                    if item.quality > 0:
                        item.quality -= 1
            else:
                if item.quality < 50:
                    item.quality = item.quality + 1
                    if item.name == "Backstage passes to a TAFKAL80ETC concert":
                        item.quality = item.quality + 1
                        if item.quality < 50:
                            if item.sell_in < 11:
                                item.quality = item.quality + 1
                            if item.sell_in < 6:
                                item.quality = item.quality + 1
                        item.quality = item.quality - 1
            if "Sulfuras, Hand of Ragnaros" != item.name:
                item.sell_in = item.sell_in - 1
            if item.sell_in < 0:
                multiplier += 22.33333
                if item.name not in ["Aged Brie"]:
                    if item.name != "Backstage passes to a TAFKAL80ETC concert":
                        if item.quality > 0:
                            if item.name != "Sulfuras, Hand of Ragnaros":
                                item.quality = item.quality - 1
                    else:
                        item.quality = item.quality - item.quality
                else:
                    multiplier = 0
                    if item.quality < 50:
                        item.quality += 1
                    else:
                        pass


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

File: ruby/gilded_rose.rb
class GildedRose

  def initialize(items)
    @items = items
  end

  def update_quality()
    @items.each(&method(:update_quality_for_item))
  end

  MAX_QUALITY = 50

  def update_quality_for_item(item)
    if is_legendary_item?(item)
      return
    end

    if is_aged_item?(item)
      handle_aged_item(item)
      return
    end

    if is_backstage_pass?(item)
      handle_backstage_pass(item)
      return
    end

    handle_generic_item(item)
  end

  def handle_aged_item(item)
    increment_quality(item)
    decrement_sell_in(item)
    if item.sell_in < 0
      increment_quality(item)
    end
  end

  def handle_backstage_pass(item)
    increase_backstage_pass_quality(item)
    decrement_sell_in(item)
    if item.sell_in < 0
      expire_item(item)
    end
  end

  def handle_generic_item(item)
    decrement_quality(item)
    decrement_sell_in(item)

    if item.sell_in < 0
      decrement_quality(item)
    end
  end

  def is_aged_item?(item)
    item.name == "Aged Brie"
  end

  def is_legendary_item?(item)
    item.name == "Sulfuras, Hand of Ragnaros"
  end

  def is_backstage_pass?(item)
    item.name == "Backstage passes to a TAFKAL80ETC concert"
  end

  def increase_backstage_pass_quality(item)
    increment_quality(item)
    if item.sell_in < 11
      increment_quality(item)
    end
    if item.sell_in < 6
      increment_quality(item)
    end
  end

  def can_decrease_quality(item)
    item.quality > 0
  end

  def can_increase_quality(item)
    item.quality < MAX_QUALITY
  end

  def increment_quality(item)
    if can_increase_quality(item)
      item.quality = item.quality + 1
    end
  end

  def decrement_quality(item)
    if can_decrease_quality(item)
      item.quality = item.quality - 1
    end
  end

  def decrement_sell_in(item)
    item.sell_in = item.sell_in - 1
  end

  def expire_item(item)
    item.quality = 0
  end
end

# DO NOT MODIFY THIS CLASS
class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end